### ISSUE TITLE
rclone copy with crypt remote, unexpected EOF, reopening on read failure

### ISSUE THREAD

**Issue by JohannesRudolph** [rclone/rclone#6451]
#### What is the problem you are having with rclone?

I have some files on an OneDrive + rclone CryptFS remote that fail to copy. The error is the same for about 5% of my files, I'm including one example that repro's this here.

#### What is your rclone version (output from `rclone version`)

```
rclone v1.59.2
- os/version: darwin 12.3.1 (64 bit)
- os/kernel: 21.4.0 (x86_64)
- os/type: darwin
- os/arch: amd64
- go/version: go1.18.6
- go/linking: dynamic
- go/tags: cmount
 ```

(I have repro'd the same on linux aarch64)

#### Which OS you are using and how many bits (e.g. Windows 7, 64 bit)
macOS (see above)

#### Which cloud storage system are you using? (e.g. Google Drive)

OneDrive, with crypt on top. This config is called `jr_rcc` in the output below

#### The command you were trying to run (e.g. `rclone copy /tmp remote:tmp`)

```
iDevBook01:Desktop jr$ rclone copy jr_rcc:Studium/Studienplan.zip ./ -vvvv
2022/09/18 21:41:03 DEBUG : rclone: Version "v1.59.2" starting with parameters ["rclone" "copy" "jr_rcc:Studium/Studienplan.zip" "./" "-vvvv"]
2022/09/18 21:41:03 DEBUG : Creating backend with remote "jr_rcc:Studium/Studienplan.zip"
2022/09/18 21:41:03 DEBUG : Using config file from "/Users/jr/.config/rclone/rclone.conf"
2022/09/18 21:41:04 DEBUG : Creating backend with remote "onedrive:jr_rcc/jr/nuokru7564aveur6l6ucho13as/elh9aunl3mm1otc62ditg8n8vs"
2022/09/18 21:41:08 DEBUG : fs cache: adding new entry for parent of "onedrive:jr_rcc/jr/nuokru7564aveur6l6ucho13as/elh9aunl3mm1otc62ditg8n8vs", "onedrive:jr_rcc/jr/nuokru7564aveur6l6ucho13as"
2022/09/18 21:41:08 DEBUG : Creating backend with remote "./"
2022/09/18 21:41:08 DEBUG : fs cache: renaming cache item "./" to be canonical "/Users/jr/Desktop"
2022/09/18 21:41:08 DEBUG : Studienplan.zip: Sizes differ (src 13349025 vs dst 6287360)
2022/09/18 21:41:13 DEBUG : Studienplan.zip: Reopening on read failure after 13303808 bytes: retry 1/10: unexpected EOF
2022/09/18 21:41:16 DEBUG : Studienplan.zip: Reopen failed after 13303808 bytes read: unexpected EOF
2022/09/18 21:41:16 NOTICE: Studienplan.zip: Removing partially written file on error: unexpected EOF
2022/09/18 21:41:16 DEBUG : Studienplan.zip: Received error: unexpected EOF - low level retry 1/10
2022/09/18 21:41:19 DEBUG : Studienplan.zip: Reopening on read failure after 13303808 bytes: retry 1/10: unexpected EOF
```

It then goes on with retries but never succeeds and eventually fails.

I tried fetching the file from OneDrive without crypt to rule out that it's an issue on OneDrive side

```
iDevBook01:Desktop jr$ rclone copy onedrive:jr_rcc/jr/nuokru7564aveur6l6ucho13as/elh9aunl3mm1otc62ditg8n8vs ./ -vvv
2022/09/18 21:42:22 DEBUG : rclone: Version "v1.59.2" starting with parameters ["rclone" "copy" "onedrive:jr_rcc/jr/nuokru7564aveur6l6ucho13as/elh9aunl3mm1otc62ditg8n8vs" "./" "-vvv"]
2022/09/18 21:42:22 DEBUG : Creating backend with remote "onedrive:jr_rcc/jr/nuokru7564aveur6l6ucho13as/elh9aunl3mm1otc62ditg8n8vs"
2022/09/18 21:42:22 DEBUG : Using config file from "/Users/jr/.config/rclone/rclone.conf"
2022/09/18 21:42:27 DEBUG : fs cache: adding new entry for parent of "onedrive:jr_rcc/jr/nuokru7564aveur6l6ucho13as/elh9aunl3mm1otc62ditg8n8vs", "onedrive:jr_rcc/jr/nuokru7564aveur6l6ucho13as"
2022/09/18 21:42:27 DEBUG : Creating backend with remote "./"
2022/09/18 21:42:27 DEBUG : fs cache: renaming cache item "./" to be canonical "/Users/jr/Desktop"
2022/09/18 21:42:28 DEBUG : elh9aunl3mm1otc62ditg8n8vs: Need to transfer - File not found at Destination
2022/09/18 21:42:31 DEBUG : elh9aunl3mm1otc62ditg8n8vs: sha1 = c08b2fffedf5e138a614d547fdd208ca5702ec4b OK
2022/09/18 21:42:31 INFO  : elh9aunl3mm1otc62ditg8n8vs: Copied (new)
2022/09/18 21:42:31 INFO  : 
Transferred:   	   12.734 MiB / 12.734 MiB, 100%, 2.331 MiB/s, ETA 0s
Transferred:            1 / 1, 100%
Elapsed time:         9.7s

2022/09/18 21:42:31 DEBUG : 9 go routines active
```

that worked out fine. Though I wouldn't know how I can manually try and verify how I can decrypt the file. 

At this point I'm thinking about ditching using crypt + OneDrive, as I have a few different configs set up that way and all of them started showing this issue on approx. 5% of files. I have backups so no problem. 

I'm 90% sure though this used to work reliably in the past as I verified integrity of my files when I initially set up that system and synced the majority of files into these remotes ~1.5 years ago. Maybe something funky has been going on OneDrive mucking with my files or a bug crept into rclone in between? 

The only other thing that may be relevant is that I used to `rclone mount` these remotes with vfs-cache quite often from different machines and maybe incomplete transfers etc. have broken them. However, there's much more files affected by this issue than I have worked on (and files I def. haven't touched) so at this point I suspect there may be a problem somewhere in rclone. 

I'd happily take any tips on how I can debug this issue further. 

#### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

**Comment by animosity22** [comment]
That generally means you have a bad copy on the remote / something didn't work out.

Are they older files? Onedrive has some data issues with the API a bit back.

**Comment by JohannesRudolph** [comment]
@animosity22 yes the files are definitely older. Do you have any link/discussion about OneDrive's API issues? 

I'd also like to verify where/how exactly the files have gone bad. Maybe it's just the metadata or sth. and could be worked around.

**Comment by animosity22** [comment]
I believe it was this thread:

https://forum.rclone.org/t/onedrive-sync-never-completes-with-sha-1-corrupted-transfer-with-latest-rclone/29884/12

@ncw or @olefrost will most likely remember much better than me as I don't use OneDrive.

**Comment by olefrost** [comment]
> https://forum.rclone.org/t/onedrive-sync-never-completes-with-sha-1-corrupted-transfer-with-latest-rclone/29884/12

The mentioned issue wouldn't corrupt any files transferred with rclone (except when explicitly disabling rclones sha-1 check),  so most likely another issue.

> I'd also like to verify where/how exactly the files have gone bad.

I suggest you start by checking the the file in the crypt against the file in your backup with these commands:

    rclone lsl jr_rcc:Studium/Studienplan.zip
    rclone hashsum sha1 jr_rcc:Studium/Studienplan.zip
    rclone hashsum sha1 jr_rcc:Studium/Studienplan.zip --download

and compare to output when the commands are performed on the file in your backup. Don't know if your backup is directly accessible by rclone, otherwise your could extract to a local testfolder and then do like this:

    rclone lsl ./testfolder/Studienplan.zip
    rclone hashsum sha1 ./testfolder/Studienplan.zip
    rclone hashsum sha1 ./testfolder/Studienplan.zip --download

If your backup is accessible with rclone, then you can also use [rclone cryptcheck](https://rclone.org/commands/rclone_cryptcheck/)

> that worked out fine. Though I wouldn't know how I can manually try and verify how I can decrypt the file.

You move the file into a new/empty folder (e.g. crypted) and then make a new crypt with the same passwords having the local folder as remote. Then you can copy it out of the folder with something like rclone copy localFolderCrypt: ./decrypted/

You can also use the hackers quick and dirty approach which is to put it into the crypt with a new name and then read it back like this (where substituting newNameCrypted with the output from the first command):

    rclone cryptdecode --reverse jr_rcc:  StudienplanFromBackup.zip
    rclone copyto ./elh9aunl3mm1otc62ditg8n8vs jr_rcc:Studium/newNameCrypted  -v
    rclone copy  jr_rcc:Studium/StudienplanFromBackup.zip . -v

**Comment by youtehub** [comment]
The same problem occurs here. After I use alist to mount BaiDu Kanbox to form WebDAV, and then use rclone to load the local hard disk, there is no problem to upload any files, but it is no problem to download files.
unexpected EOF will appear when a file is downloaded. The contents are as follows:
2022/10/09 12:48:47 ERROR : IO error: error when trying to read error from body: unexpected EOF

**Comment by youtehub** [comment]
I really want to understand what causes, is the time to download, if the local cache without this folder, it cannot automatically create folder, because according to the execution of the program, will first check whether there is a local cache files with the same name, is picked up directly from the local cache copy, if not, first downloaded to the local cache, Then copy it from the local cache to the destination folder.
Feel from the Baidu web disk to read the file, the file is written to the cache when the problem, it seems that there is no go through the above process.
It feels like there is a problem when you download from webdav to local cache and then from cache to destination folder if there is no target file in the local cache.

**Comment by animosity22** [comment]
@JohannesRudolph  still an issue? have questions?

**Comment by JohannesRudolph** [comment]
@animosity22 and @olefrost thanks for all the detailed input. I wasn't able to 100% pinpoint the root cause for my issue other than that the files are definitely corrupt on OneDrive. Whether they got corrupted on rclone upload side or on OneDrive API side I can't say for certain anymore from what records I have.

What I took away from this is that having too many moving parts in my file replication chain was not a good idea and thus simplified my setup considerably. Still use rclone + crypt on OneDrive for some of my files and after freshly copying over so far everything checks out. I have replaced some of my backups uses cases of rclone + crypt with kopia + rclone (kopia does encryption so I figured I can skip crypt).

Since this issue is not really reproducible and actionable for you guys, I'm closing it. 

Thanks for all the help!
